### PR TITLE
net: sockets_service: Fix thread failure

### LINKED ISSUE #73523: net: sockets_service: Registering a service in case of socket service init failure hangs
**Describe the bug**
In case of sockets service init failure, if any module attempts to register a socket then it leads to hang. This is due to usage of conditional variable, in `z_impl_net_socket_service_register` if the init is not done then it waits on a conditional variable forever `k_condvar_wait(&wait_start, &lock, K_FOREVER);`, the assumption here is that during boot init is still in progress and eventually once init is completed (pass/fail) `socket_service_thread` calls `k_condvar_broadcast(&wait_start);`.

But in case of any negative cases e.g., `CONFIG_NET_SOCKETS_POLL_MAX` is less than number of services, then `socket_service_thread` fails and release the conditional variable, and then if any service registers then it waits for conditional variable to be released but no one to release.

**To Reproduce**
1. `CONFIG_NET_SOCKETS_POLL_MAX` to a lower value e.g., 4, you will see a warning in boot
2. try `net dhcpv4 server start 1 192.168.1.2` this tries to register a socket and hangs.

**Expected behavior**
Registration can fail, but shouldn't hang.

**Impact**
I would consider this as a major bug in case of misconfiguration.

**Logs and console output**
NA

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain (e.g Zephyr SDK, ...):  Zephyr SDK
 - Commit SHA or Version used: 3537c4614bc

**Additional context**
NA

### PR DESCRIPTION
**Pull request by krish2718**
In case of thread failure, fix the registrations by properly managing the synchronization i.e. use conditional wait only if thread is being initialized or will be initialized, else check for success or failure without waiting for conditional variables.

Fixes #73523.

With this fix see below graceful failure instead of hang.
```
*** Booting nRF Connect SDK v2.6.99-cef9952e037f ***
*** Using Zephyr OS v3.6.99-80ffee125b69 ***
Starting nrf7002dk with CPU frequency: 128 MHz
[00:00:00.613,494] <inf> net_config: Initializing network
[00:00:00.613,494] <inf> net_config: Waiting interface 1 (0x20001388) to be up...
[00:00:00.613,555] <inf> net_config: IPv4 address: 192.168.1.1
[00:00:00.614,349] <inf> wpa_supp: Successfully initialized wpa_supplicant
[00:00:00.616,455] <err> net_sock_svc: You have 12 services to monitor but 10 poll entries configured.
[00:00:00.616,485] <err> net_sock_svc: Please increase value of CONFIG_NET_SOCKETS_POLL_MAX to at least 12
uart:~$ net dhcpv4 server start 1 192.168.1.2
DHCPv4 server failed to start on interface 1, error 5
[00:00:13.379,821] <err> net_dhcpv4_server: Failed to register socket service, -5
```

### CODE REVIEW

**Comment by jukkar** [comment]
Thanks @krish2718, looks good. FYI, I have a PR #73422 pending where the socket_service thread is started differently, basically that commit in the PR could also be taken separately and it would probably fix the issue too.

**Comment by krish2718** [comment]
> Thanks @krish2718, looks good. FYI, I have a PR #73422 pending where the socket_service thread is started differently, basically that commit in the PR could also be taken separately and it would probably fix the issue too.

Based on a quick look, IIUC, we need this fix for that PR as well, as the conditional wait logic hasn't changed.

**Comment by jukkar** [comment]
> Based on a quick look, IIUC, we need this fix for that PR as well, as the conditional wait logic hasn't changed.

I was pondering the problem described in this PR a bit more, and I am worried that we might hide an error here and let the system "run" without it actually being able to work properly. What I mean is that if user has not enabled enough polled entries in the system, and we do notice it at runtime, so this message
```
[00:00:00.616,455] <err> net_sock_svc: You have 12 services to monitor but 10 poll entries configured.
[00:00:00.616,485] <err> net_sock_svc: Please increase value of CONFIG_NET_SOCKETS_POLL_MAX to at least 12
```
we should probably assert here so that the situation can be fixed. If we return an error instead of waiting but not actually fix the issue, the device might not work properly but the user might not realize it. So it might be better to just assert if this configuration issue is seen instead of continuing. @rlubos / @krish2718 WDYT?

**Comment by krish2718** [comment]
> So it might be better to just assert if this configuration issue is seen instead of continuing

But the problem is Assert's are not enabled by default except in test, so, this won't be useful to the user and we still need to handle the registration based on init status.

File: subsys/net/lib/sockets/sockets_service.c [hunk 1]
```diff
@@ -63,8 +69,11 @@ int z_impl_net_socket_service_register(const struct net_socket_service_desc *svc
 
 	k_mutex_lock(&lock, K_FOREVER);
 
-	if (!init_done) {
+	if (thread_status == SOCKET_SERVICE_THREAD_UNINITIALIZED) {
 		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	} else if (thread_status != SOCKET_SERVICE_THREAD_RUNNING) {
+		ret = -EIO;
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets_service.c:75]: I'd print some error here.

**Review by rlubos — DISMISSED** [review]

**Comment by jukkar** [comment]
If absolutely necessary this could be accepted, it just needs immediate fixes again. It is more or less a 10 min job to add one CHECKIF call for additional checks.

**Comment by jukkar** [comment]
I added an error print when the register fails so we get some indication to the user to figure out what to do next.

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**